bin/generate-zbm

```diff
@@ -523,22 +526,91 @@ sub versionedKernel {
   return;
 }
 
-# Finds the latest kernel in /boot, based on form <prefix>-<version>
+# Finds the latest kernel in /boot, if possible
 sub latestKernel {
-  my @prefixes = ( "vmlinux*-*", "vmlinuz*-*", "linux*-*", "kernel*-*" );
+  my @prefixes = ( "vmlinuz*", "vmlinux*", "linux*", "kernel*" );
+
   for my $prefix (@prefixes) {
-    my $glob    = join( '/', ( $runConf{bootdir}, $prefix ) );
-    my @kernels = glob($glob);
-    next if !@kernels;
-    for ( sort { versioncmp( $b, $a ) } @kernels ) {
+    my $glob = join( '/', ( $runConf{bootdir}, $prefix ) );
+    my %kernels;
+
+    for my $kernel ( glob($glob) ) {
+      my $version = kernelVersion($kernel);
+      next unless defined($version);
+      Log("Identified version $version for kernel $kernel");
+      $kernels{$version} = $kernel;
+    }
+
+    next unless ( keys %kernels );
+
+    for ( sort { versioncmp( $b, $a ) } keys %kernels ) {
       Log("Latest kernel: $_");
-      return $_;
+      return $kernels{$_};
     }
   }
 
   return;
 }
 
+# Attempts to determine a version for the given kernel, by
+#
+# a. Identifying the first version-looking string in the file, or
+# b. Identifying a version-like part in the name of the file
+#
+# If one of these exists and not the other, that value is used;
+# if both exist, they must agree or nothing is returned
+sub kernelVersion {
+  my $kernel = shift;
+
+  my ( $filever, $namever );
+
+  # Consider an unreadable file to have no version
+  unless ( -r $kernel ) {
+    Log("Unable to read path $kernel, assuming no version");
+    return;
+  }
+
+  # Read strings in the kernel to recover a version, if possible
+  my @output = execute(qq(strings $kernel));
+  my $status = pop(@output);
+  if ( $status eq 0 ) {
+    for (@output) {
+      if (/[0-9]+\.[0-9]+\.[0-9]+/) {
```

Comment:
Should this be anchored to the start of a line, to avoid possibly matching against a random string in the middle of gibberish?